protocols/randhound/randhound.go

```diff
@@ -1,251 +1,995 @@
+// Start initiates the RandHound protocol run. The client pseudo-randomly
+// chooses the server grouping, forms an I1 message for each group, and sends
+// it to all servers of that group.
+func (rh *RandHound) Start() error {
+
+	var err error
+
+	// Set timestamp
+	rh.time = time.Now()
+
+	// Choose client randomness
+	hs := rh.Suite().Hash().Size()
+	rand := make([]byte, hs)
+	random.Stream.XORKeyStream(rand, rand)
+	rh.cliRand = rand
+
+	// Determine server grouping
+	rh.server, rh.key, err = rh.Shard(rand, rh.groups)
+	if err != nil {
+		return err
+	}
+
+	// Set some group parameters
+	for i, group := range rh.server {
+		rh.threshold[i] = 2 * len(group) / 3
+		rh.polyCommit[i] = make([]abstract.Point, len(group))
+		g := make([]int, len(group))
+		for j, server0 := range group {
+			s0 := server0.ServerIdentityIdx
+			rh.ServerIdxToGroupNum[s0] = i
+			rh.ServerIdxToGroupIdx[s0] = j
+			g[j] = s0
+		}
+		rh.group[i] = g
+	}
+
+	// Compute session id
+	rh.sid, err = rh.sessionID(rh.nodes, rh.faulty, rh.purpose, rh.time, rh.cliRand, rh.threshold, rh.Public(), rh.key)
+	if err != nil {
+		return err
+	}
+
+	// Multicast first message to grouped servers
+	for i, group := range rh.server {
+
+		index := make([]uint32, len(group))
+		for j, server := range group {
+			index[j] = uint32(server.ServerIdentityIdx)
+		}
+
+		i1 := &I1{
+			SID:       rh.sid,
+			Threshold: rh.threshold[i],
+			Group:     index,
+			Key:       rh.key[i],
+		}
+
+		rh.mutex.Lock()
+
+		// Sign I1 and store signature in i1.Sig
+		if err := signSchnorr(rh.Suite(), rh.Private(), i1); err != nil {
+			return err
+		}
+
+		rh.i1s[i] = i1
+
+		rh.mutex.Unlock()
+
+		if err := rh.Multicast(i1, group...); err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
-// Peer (=server) refers to a node which contributes to the generation of the
-// public randomness.
-type Peer struct {
-	rs     []byte              // A peer's trustee-selection random value
-	shares map[uint32]*R4Share // A peer's shares
-	i1     *I1                 // I1 message we received from the leader
-	i2     *I2                 // I2 - " -
-	i3     *I3                 // I3 - " -
-	i4     *I4                 // I4 - " -
-	r1     *R1                 // R1 message we sent to the leader
-	r2     *R2                 // R2 - " -
-	r3     *R3                 // R3 - " -
-	r4     *R4                 // R4 - " -
+// Shard produces a pseudorandom sharding of the network entity list
+// based on a seed and a number of requested shards.
+func (rh *RandHound) Shard(seed []byte, shards int) ([][]*sda.TreeNode, [][]abstract.Point, error) {
+
+	if shards == 0 || rh.nodes < shards {
+		return nil, nil, fmt.Errorf("Number of requested shards not supported")
+	}
+
+	// Compute a random permutation of [1,n-1]
+	prng := rh.Suite().Cipher(seed)
+	m := make([]int, rh.nodes-1)
+	for i := range m {
+		j := int(random.Uint64(prng) % uint64(i+1))
+		m[i] = m[j]
+		m[j] = i + 1
```

Comment:
It's basically an adaptation of Golang's `rand.Perm(...)` function, see here: https://golang.org/src/math/rand/rand.go?s=5071:5103#L164